### ISSUE TITLE
Provide a "bom" for Maven Xtext dependencies

### ISSUE THREAD

**Issue by LorenzoBettini** [eclipse/xtext-maven#35]
I was wondering if we should provide a "bom" for Maven dependencies used in Xtext Maven projects; this should only list dependencies in _dependencyManagement_ section with the versions, and Maven projects could use something like

```
<dependencyManagement>
    <dependencies>
        <dependency>
            <groupId>org.eclipse.xtext</groupId>
            <artifactId>org.eclipse.xtext.maven.bom</artifactId>
            <version>XTEXT VERSION</version>
            <type>pom</type>
            <scope>import</scope>
        </dependency>
    </dependencies>
</dependencyManagement>
```

This way, actual dependency versions should not be kept updated throughout all the Maven projects each time a new version is released, but only in the bom.

Not to mention that versions are hardcoded in poms generated by our wizards, like in the .ide project:

```
	<dependencies>
		<dependency>
			<groupId>log4j</groupId>
			<artifactId>log4j</artifactId>
			<version>1.2.16</version>
		</dependency>
		<dependency>
			<groupId>org.eclipse.lsp4j</groupId>
			<artifactId>org.eclipse.lsp4j</artifactId>
			<version>0.2.1</version>
		</dependency>
		<dependency>
			<groupId>org.ow2.asm</groupId>
			<artifactId>asm</artifactId>
			<version>5.0.1</version>
		</dependency>
		<dependency>
			<groupId>org.ow2.asm</groupId>
			<artifactId>asm-commons</artifactId>
			<version>5.0.1</version>
		</dependency>
		<dependency>
			<groupId>org.ow2.asm</groupId>
			<artifactId>asm-tree</artifactId>
			<version>5.0.1</version>
		</dependency>
	</dependencies>
```

and using deps from the same groupId forces to repeat the version consistently (like in the ow2.asm above).

The approach of the bom seems to be a success in projects like Spring and Spring Boot, for example.

**Comment by LorenzoBettini** [comment]
@cdietrich @dhuebner what do you think?

**Comment by cdietrich** [comment]
may maven knowledge is too bad to know which possible negative side effect this has. and i wonder if we would need the same at other places. e.g. the maven stuff the wizard creates. @kthoms what do you think

**Comment by kthoms** [comment]
I think this is a good approach. I was thinking also in this direction for providing parent POMs.

**Comment by LorenzoBettini** [comment]
@cdietrich  The idea is that the wizard should create such a section in the parent pom of the created project:

```
<dependencyManagement>
    <dependencies>
        <dependency>
            <groupId>org.eclipse.xtext</groupId>
            <artifactId>org.eclipse.xtext.maven.bom</artifactId>
            <version>XTEXT VERSION</version>
            <type>pom</type>
            <scope>import</scope>
        </dependency>
    </dependencies>
</dependencyManagement>
```

and then, for example for the .ide project, generate dependencies without version, e.g., (compare this to what happens now)

```
	<dependencies>
		<dependency>
			<groupId>log4j</groupId>
			<artifactId>log4j</artifactId>
		</dependency>
		<dependency>
			<groupId>org.eclipse.lsp4j</groupId>
			<artifactId>org.eclipse.lsp4j</artifactId>
		</dependency>
		<dependency>
			<groupId>org.ow2.asm</groupId>
			<artifactId>asm</artifactId>
		</dependency>
		<dependency>
			<groupId>org.ow2.asm</groupId>
			<artifactId>asm-commons</artifactId>
		</dependency>
		<dependency>
			<groupId>org.ow2.asm</groupId>
			<artifactId>asm-tree</artifactId>
		</dependency>
	</dependencies>
```

The versions are taken from the bom Maven project (which should list deps and versions in a dependencyManagement section, so that versions will be effective only when specified in the poms using the bom).

When a new version of Xtext is going to be released, possible version changes will be stored in the bom.

The projects using the bom only need to change the Xtext version and the new versions of deps will be automatically used.

@kthoms using the bom has the advantage that it can be used also in projects where the parent pom cannot be changed.

Spring boot provides both a parent pom and a bom in case you want to use your own parent.

Of course the parent pom would enable also the default configuration of Maven plugins (in a pluginManagement section) which would be good as well :)

I think the standard approach is to have a parent pom which in turns has the bom as the parent. This way you can use the parent pom or the bom consistently.

For example, spring-boot-starter-parent has as parent spring-boot-dependencies (which is the bom).

**Comment by cdietrich** [comment]
but that would not allow to have different deps at different places correct?

**Comment by LorenzoBettini** [comment]
@cdietrich summarizing, the Xtext wizards should not be changed when versions of dependencies change (just use the new version of Xtext in the generated poms, but it's just a thing to change); of course if new dependencies are needed in the poms generated by the wizards should be changed as well.

Of course, poms using the bom are allowed to override versions of dependencies.

In general boms are used to make sure that the dependencies are used with the same versions as the one used by the framework itself, but if you want to override them you're allowed to do so.

**Comment by cdietrich** [comment]
build calls 
mvn -f git-repo/maven/org.eclipse.emf.mwe2.parent/pom.xml -s /opt/public/hipp/homes/genie.xtext/.m2/settings-deploy-ossrh.xml -DBUILD_TYPE=N -Dsign=true -DBUILD_ALIAS= -Dsign.skip=false -DtestFailureIgnore=true -Dmaven.repo.local=/home/hudson/genie.xtext/.jenkins/jobs/mwe2/workspace/.repository clean install

=> @kthoms is is possible jenkinsfile not used?

**Comment by LorenzoBettini** [comment]
@cdietrich @kthoms I may start working on this, do you think it's worthwhile?
it will then require changes in the wizards.

**Comment by cdietrich** [comment]
For me it would be ok @kthoms what do you think

**Comment by kthoms** [comment]
Can we consider this as fixed with introduction of the dev-bom with eclipse/xtext#1249 ?

**Comment by cdietrich** [comment]
i assume @LorenzoBettini was thinking a bom including xtext plugins as well

**Comment by kthoms** [comment]
Ah, yes, this is still an open issue.

**Comment by LorenzoBettini** [comment]
Well, if I understand correctly, the bom, which only deals with dependencies, should be done now, shouldn't it? For plugins you need a parent POM, you can't do that with a bom. Or are we talking about the dependencies of plugins themselves?

**Comment by kthoms** [comment]
Xtext's Maven consumable artifacts, e.g. everything from xtext-lib/core/extras, could be managed in a BOM, too. The new xtext-dev-bom only manages dependencies Xtext itself has.
So if you would like to have xtext artifacts managed by a BOM, then this is missing now.

**Comment by LorenzoBettini** [comment]
OK, so for the moment external deps are not managed by the BOM, are they?

**Comment by cdietrich** [comment]
on master they are. you are strongly encouraged to test

**Comment by cdietrich** [comment]
https://www.eclipse.org/forums/index.php/t/1097108/

**Comment by LorenzoBettini** [comment]
Now that the BOM has been released, I see that the BOM does not specify in the dependencyManagement section all the Xtext Maven artifacts... why is that? Moreover, the generated POM in the web project has all dependencies with the versions hardcoded, thus these dependencies are not in the BOM either?

Is that intentional? @cdietrich @kthoms

**Comment by cdietrich** [comment]
@LorenzoBettini i cannot follow you. please be more explicit

**Comment by cdietrich** [comment]
https://ci.eclipse.org/xtext/job/xtext-core/job/master/lastSuccessfulBuild/artifact/build/maven-repository/org/eclipse/xtext/org.eclipse.xtext.ide/2.18.0-SNAPSHOT/org.eclipse.xtext.ide-2.18.0-20190326.230423-1.pom/*view*/

**Comment by LorenzoBettini** [comment]
When a project provides a BOM, besides the dependencies used by the project's modules should also list all the modules of the project in the BOM itself.

Thus, the BOM should also contain references to all the Xtext Maven artifacts, with the corresponding versions. For example, it should list

```xml
<dependency>
	<groupId>${project.groupId}</groupId>
	<artifactId>org.eclipse.xtext.xbase.lib</artifactId>
	<version>${project.version}</version>
</dependency>
``` 

So that clients, once imported the BOM, can simply specify as dependencies

```xml
<dependency>
	<groupId>org.eclipse.xtext</groupId>
	<artifactId>org.eclipse.xtext.xbase.lib</artifactId>
</dependency>
```

Moreover, the web project generated by the wizard lists all the dependencies with hardcoded versions, e.g.,

```xml
<dependency>
	<groupId>org.webjars</groupId>
	<artifactId>requirejs</artifactId>
	<version>2.3.6</version>
</dependency>
```

While these should be specified in the BOM as well (I mean their versions)

**Comment by cdietrich** [comment]
and for the webpart:
it might have been intentiopn not to include this in the bom
since it is used in the web part only

**Comment by cdietrich** [comment]
i dont know if there are any best practices for that
a framework that has a shitload of deps but that may be relevant only for a minimal part / special usecase

**Comment by LorenzoBettini** [comment]
You don't need to know the version in advance, that's why in the BOM you should specify something like

```xml
<dependency>
	<groupId>${project.groupId}</groupId>
	<artifactId>org.eclipse.xtext.xbase.lib</artifactId>
	<version>${project.version}</version>
</dependency>
```

The Xtext Maven artifacts will be specified with the same version of the BOM :)

Can I work on that?

I think for the web project it was just forgotten, and it makes sense to list the web project's dependencies in the BOM...

Also the IdeProjectDescriptor does not use the BOM: https://github.com/eclipse/xtext-core/blob/1aea0c1928a8512548a85f8e8bff3f1f4918f6f5/org.eclipse.xtext.xtext.wizard/src/org/eclipse/xtext/xtext/wizard/IdeProjectDescriptor.xtend#L48 while again it should, in my opinion...

**Comment by cdietrich** [comment]
i want this to be discussed with @szarnekow and @kthoms first.
unfortunately we are all busy like hell right know
me answering you give me overtime already

**Comment by cdietrich** [comment]
@LorenzoBettini doesnt ide inherit from parent
see https://github.com/eclipse/xtext-core/issues/1076
where again @szarnekow and @kthoms did not look at yet

**Comment by LorenzoBettini** [comment]
OK, no problem, I can work on that if you think it's worthwhile.
Summarizing, the BOM should provide a curated list of dependencies, including all the modules of the project itself. For example, log4j bom lists all its modules: https://github.com/apache/logging-log4j2/blob/master/log4j-bom/pom.xml and the Spring Boot BOM lists its modules and all the dependencies (including test dependencies, but that would not be relevant in our context).

Each time a new version of Xtext (BOM) is released all clients using the BOM will automatically use the updated dependencies (if versions are not hardcoded in the POM, as it is now for the web project and the ide project). Currently, the versions of dependencies (e.g., of the web project) must be manually updated in our wizard and the clients will have to regenerate the POM, which breaks most of the utility and goals of the BOM

**Comment by cdietrich** [comment]
i am not sure if this will explode. none of the downstream stuff is available when we built the bom
or when we build xtext-core using the bom

**Comment by cdietrich** [comment]
we also are i a kind of in between state regarding the bom cause
https://github.com/eclipse/xtext/issues/1388

**Comment by cdietrich** [comment]
=> i currenty have no idea if this is a good idea and if it is worth working on this
(i am not a maven nor a gradle wizard, but i fear nobody actively working on Xtext is and the feedback from the community is very poor)

**Comment by LorenzoBettini** [comment]
I still don't understand this part

"i am not sure if this will explode. none of the downstream stuff is available when we built the bom
or when we build xtext-core using the bom"

In any case, before releasing a new version of Xtext, you need to manually change the versions in the wizard, thus, at that time, you simply change the versions in the BOM, it's just the same, but with the advantages I described above...

From my understanding, please correct me if I'm wrong, the current BOM solved many problems due to JDT and Eclipse platform version ranges in the xtend-maven-plugin and xtext-maven-plugin right? But that's all (don't get me wrong, it's a really good thing ;) but in the current state, clients benefit from these advantages even without importing the bom in their parent pom. The consistencies of dependencies are guaranteed by the bom for xtext maven plugins. But not for the dependencies used by the clients' poms.

In fact, the ide and the web project still specify all versions in an hardcoded way.

Summarizing, you still need the effort of updating dependencies in the wizard (without real benefits for the users, since their poms are not updated, unless they regenerate them). It's just a matter of updating the versions in the bom instead of in the wizard, with huge benefits for the users.

**Comment by cdietrich** [comment]
i am not talking about xtext client projects.
i am talking about building

xtext-lib, xtext-core, xtext-extras, xtext-web, xtext-xtend with gradle
xtext-eclipse and xtext-xtend with tycho
xtext-maven and xtext-xtend with pure maven

**Comment by cdietrich** [comment]
and no, there is no guaranty. it depends on where in the level xtext and as where others are

**Comment by cdietrich** [comment]
and web and idea dont specify emf for example

**Comment by cdietrich** [comment]
but as i said: i am NOT a maven wizard.
i see only what we have a `<insert curses here>` repo split with `<insert curses here>`  3 different kinds of builds on 9 repos and i want to be sure this works.

**Comment by LorenzoBettini** [comment]
Then probably I'm missing something in the current build process I'm afraid...
Or probably I'm not able to explain myself...

and also the sentence of web and ide that don't specify emf is mysterious to me ;)
what's imported with the BOM is not automatically a dependency of a user...
if the user (importing the bom) specifies that dependency then it will be automatically used with the imported version, that's all.

A BOM with 100 dependencies in the dependencyManagement section does not mean that the user has 100 dependencies :)

**Comment by cdietrich** [comment]
yes. but i simply want to make sure it works. so it would be up to you to do so.
and it still should work if we do eclipse/xtext#1388

**Comment by cdietrich** [comment]
and again: this is not about the user. it is about building xtext itself

**Comment by cdietrich** [comment]
and again: i dont know if the versions in ide and web are intentional, merge errors or were forgotten

**Comment by LorenzoBettini** [comment]
OK, I give up, please ping me if I can provide help ;)

**Comment by LorenzoBettini** [comment]
According to https://blogs.itemis.com/en/eclipse-xtext-goes-shopping-the-new-bill-of-materials versions in the ide and web are simply forgotten

> So, clients do no longer carefully revise each and every dependency, which thus lowers maintenance efforts. Also, clients can be sure that the managed versions of all libraries have been tested to work properly with Xtext and with each other.

;)

**Comment by cdietrich** [comment]
@LorenzoBettini please dont
lets wait for @szarnekow and @kthoms

**Comment by LorenzoBettini** [comment]
Yes, of course I meant "I give up for the moment", not "for good" ;)

and as a further note (to keep it in the discussion) you can safely name a dependency in the BOM even if it does not exist; I guess that's related to the building process, if I understand it correctly

**Comment by cdietrich** [comment]
yes i am not sure if we can do what if we create it with gradle
you might have a look at my 
`cd_javaPlatformExperiments`
branches in lib and core

**Comment by LorenzoBettini** [comment]
Well, in that case, I probably won't be able to help... I really hate gradle and I'm currently not planning to ever deal with that ;) I have invested much on Maven, which works like a charm, it's a build system and behaves like a build system, not like a programming language :)

jokes apart, I'm still available for the Maven side

**Comment by kthoms** [comment]
We have put this on our list for discussion in next week and let you know the results here.

**Comment by cdietrich** [comment]
clear milestone
